pulumi_state_splitter: initial PR

filip-zyzniewski opened the pull request.

filip-zyzniewski committed:
pulumi_state_splitter: initial boilerplate

filip-zyzniewski committed:
pulumi_state_splitter: check tests code coverage

filip-zyzniewski committed:
pulumi_state_splitter: tests: add a basic Pulumi smoke test

filip-zyzniewski committed:
pulumi_state_splitter: introduce the state_file module

filip-zyzniewski committed:
pulumi_state_splitter: initial working implementation

filip-zyzniewski committed:
pulumi_state_splitter: implement the command line interface

filip-zyzniewski committed:
pulumi_state_splitter: cli: prepare for adding the `run` command

I went for consistency between different commands.

`--all` => `--all-stacks`
  `run --all-stacks` seems clearer than `run --all`

`stack1 stack2` => `--stack stack1 --stack stack2`
 to avoid hitting
 https://github.com/pallets/click/blob/cab9483a30379f9b8e3ddb72d5a4e88f88d517b6/src/click/parser.py#L96
 for the `run` command with stacks and command arguments.

filip-zyzniewski committed:
pulumi_state_splitter: implement split.Unsplitter

filip-zyzniewski committed:
pulumi_state_splitter: cli: implement the `run` command

filip-zyzniewski committed:
pulumi_state_splitter: add a github workflow for tests

filip-zyzniewski committed:
pulumi_state_splitter: cli: implememt StackNameOptionType

This gives us a nicer --help output:

was:
  `-s, --stack TEXT`
is:
  `-s, --stack PROJECT-NAME/STACK-NAME`

filip-zyzniewski committed:
pulumi_state_splitter: extend the README

filip-zyzniewski committed:
pulumi_state_splitter: expect 100% test coverage

Let's try not making it worse than it is

pbzdyl commented on .gitignore:
**Suggestion**: these ignores seem to be specific to `pulumi_state_splitter` so maybe we should keep them in `utilities/pulumi_state_splitter/.gitignore`?

filip-zyzniewski replied:
Partially done in 6f4983cb5e31d6702960853f0c0d3778c1ffda39, `__pycache__` and `.coverage` make sense for any Python project, so I have left them in the root one.

pbzdyl commented on .github/workflows/utilities-pulumi_state_splitter.yml:
```diff
@@ -0,0 +1,15 @@
+      - uses: actions/checkout@master
+      - run: >
+          sudo apt update &&
+          sudo apt install python3.11
```
**Suggestion**: you could consider using https://github.com/actions/setup-python instead as it can also setup dependencies cache for the builds and setup [problem matchers](https://github.com/actions/toolkit/blob/main/docs/problem-matchers.md) for [python specific console output](https://github.com/actions/setup-python/blob/main/.github/python.json).

filip-zyzniewski replied:
Done in c155b6a707ad0b427340bad05612daea121b9bcc and 60090109501820d87e8d86facb652817109730bf .

pbzdyl commented on utilities/pulumi_state_splitter/pulumi_state_splitter/cli.py:
```diff
@@ -0,0 +1,124 @@
+    @functools.wraps(f)
+    def check_arguments(backend_dir, all_stacks, stack, **kwargs):
+        if all_stacks and stack:
+            raise ValueError("--all-stacks is mutually exclusive with --stack")
```
Alternatively you could use a "magic" value for `--stack` (e.g. `*`) meaning "all stacks" or assume all stacks when `stack` is not provided and have only one option instead of two mutually exclusive ones.

filip-zyzniewski replied:
I am not a big fan of magic values.
I wanted the user to be specific about what they want to happen.
I can reconsider if you think that the current approach is a bad idea.

pbzdyl replied:
Yep, magic value sounds like a bad idea. On the other hand assuming all stack if you don't specify any might be a good default and it will simplify the code a lot (no more tracking of two different options - just find/discover all stacks if none has been provided). I'll leave this up to you.

filip-zyzniewski replied:
Not sure, but went for it in 2bda01a14de0eb583b20c9ce96dbc41a1fae322c .

pbzdyl commented on utilities/pulumi_state_splitter/pulumi_state_splitter/model.py:
```diff
@@ -0,0 +1,93 @@
+        data = handler(self)
+        if not data["latest"]:
+            data.pop("latest")
+        return data
```
Did you consider extracting a common function to handle omitting falsey attributes in `Resource` and `Checkpoint`?

filip-zyzniewski replied:
Done in 0fe4bc2ea54b0673d2c0113b46f45e55d511a2e0 .

pbzdyl commented on utilities/pulumi_state_splitter/pulumi_state_splitter/model.py:
```diff
@@ -0,0 +1,93 @@
+    }
+
+    checkpoint: Checkpoint
+    version: int
```
I'm guessing this tool requires a specific version of the state file format. Shall we warn or even fail when the actual version found in the state file is different than the version supported here (3?)?

filip-zyzniewski replied:
Done in d77be278fd39f38b8c4d040efa145912c8e228a8

pbzdyl commented on utilities/pulumi_state_splitter/pulumi_state_splitter/state_file.py:
```diff
@@ -0,0 +1,94 @@
+def sorted_resources(
+    resources: Iterable[pulumi_state_splitter.model.Resource],
+) -> List[pulumi_state_splitter.model.Resource]:
+    """Sorts resources to Pulumi's liking."""
```
**Suggestion**: you could elaborate more what it actually means :)

filip-zyzniewski replied:
Done in 1ecd9a749fe7b0c2f505d6000a292cd46dc616c4 .

pbzdyl commented on utilities/pulumi_state_splitter/pulumi_state_splitter/state_file.py:
```diff
@@ -0,0 +1,94 @@
+    # sort by URN first
+    urn2resource = dict(sorted((resource.urn, resource) for resource in resources))
+
+    output = {}
```
It took me a while to figure out why you are using a dict here and that you rely on the insertion order being preserved by dicts so maybe you could add a comment explaining that context for future readers.

pbzdyl replied:
OK, now that I got back to this function again, it is now clearer to me, so I guess I just needed to internalize the properties of python dict :)

pbzdyl commented on utilities/pulumi_state_splitter/pulumi_state_splitter/state_file.py:
```diff
@@ -0,0 +1,94 @@
+        cls, backend_dir: pathlib.Path
+    ) -> Iterable[pulumi_state_splitter.stored_state.StackName]:
+        """Finds all file states in the Pulumi backend directory."""
+        glob_states = cls._glob_all().path
```
This calls `_glob_all()` from the superclass and then calls `path` that is not defined on the superclass - it's defined on the subclasses though. Should we have an abstract `path` method defined in `StoredState` class?

filip-zyzniewski replied:
The "trick" here is that we are making a glob pointing at a particular YAML file (`project-name/stack-name/state.yaml`), not the whole StateDir. 
Unlike StateFile, which is looking up the JSON files (so the glob path matches the "state" path).

pbzdyl commented on utilities/pulumi_state_splitter/pulumi_state_splitter/state_file.py:
```diff
@@ -0,0 +1,94 @@
+            self.path.parent,
+            self.path.parent.parent,
+            self.path.parent.parent.parent,
+        ):
```
Am I reading it correctly (per https://github.com/house-reliability-engineering/pulumi/pull/1/files?short_path=dacdb4b#diff-6d8fbaa4c599037030cdfc92f69840df2484fcfa2aa129178745f26d01be5b35R22-R28) that organization part in the directory hierarchy is optional? If yes, is the number of partent directories we are removing correct in all cases?

filip-zyzniewski replied:
Nope, the organization part is only supported for `StackName`, and since Pulumi won't accept any other organization than `organization` for `file://` backends, we do not put the organization name in the paths. It would have to always be `organization`.

pbzdyl commented on utilities/pulumi_state_splitter/pulumi_state_splitter/state_file.py:
```diff
@@ -0,0 +1,94 @@
+    def load(self):
+        """Loads the contents of the state file."""
+        self.state = pulumi_state_splitter.model.State.model_validate_json(
+            self.path.read_text()
```
Noob's question: this will read the whole JSON text into memory and then parse it. This probably won't be a problem for pulumi state files as they are not that big, but is it even possible to stream the data from the file and parse it without loading its whole text content into memory?

filip-zyzniewski replied:
We could try something like https://pypi.org/project/json-stream/, but that would require changing our APIs to support streaming loading of the state models.
I think it would be better to consider this if we actually have a problem with the size of the JSON.

pbzdyl replied:
It's not a concern for me, I was just curious as other APIs like json or yaml packages support this out of the box.

filip-zyzniewski committed:
pulumi_state_splitter.cli: add missing imports

pbzdyl commented on utilities/pulumi_state_splitter/pulumi_state_splitter/split.py:
```diff
@@ -0,0 +1,177 @@
+                        resource.outputs = yaml.load(f, yaml.Loader)
+                resources.append(resource)
+        self.state.checkpoint.latest.resources = (
+            pulumi_state_splitter.model.Resource.find_parents(resources)
```
Am I correct that we want to have that invariant that the resources have always their parents set? Maybe we could have a custom "setter" method for setting `Latest.resources` that would always set its value after passing it through `Resource.find_parents`?

filip-zyzniewski replied:
Not necessarily. The stack and the providers usually don't have a parent. I might be misunderstanding your question though.

filip-zyzniewski replied:
Hm, not sure about the setter idea, as sometimes we do sometimes get `parent_resource`, for example @ 
https://github.com/house-reliability-engineering/pulumi/blob/1ecd9a749fe7b0c2f505d6000a292cd46dc616c4/utilities/pulumi_state_splitter/pulumi_state_splitter/split.py#L31
or
https://github.com/house-reliability-engineering/pulumi/blob/1ecd9a749fe7b0c2f505d6000a292cd46dc616c4/utilities/pulumi_state_splitter/pulumi_state_splitter/split.py#L35

WDYT?

pbzdyl replied:
Ah, ok, we there are cases where there is no point in calling that method as the parents information is already populated.

pbzdyl commented on utilities/pulumi_state_splitter/pulumi_state_splitter/split.py:
```diff
@@ -0,0 +1,177 @@
+                    data = yaml.load(f, yaml.Loader)
+                resource = pulumi_state_splitter.model.Resource.model_validate(data)
+                if resource.type == "pulumi:pulumi:Stack":
+                    with (self.path / "outputs.yaml").open() as f:
```
IIRC normal (not stack) resources also might have outputs. What's the reason for treating the stack resource outputs differently?

filip-zyzniewski replied:
I wanted to have the stack's outputs easily available at `project/stack/outputs.yaml`.
Resource outputs are not that interesting on their own and there is a multitude of them - we'd effectively double the amount of YAML files if we did that for resource outputs as well.

pbzdyl replied:
I meant exactly that - why bother with special handling for stack outputs and not keep them inline with the stack resource yaml :) But current approach works for me too.

filip-zyzniewski replied:
Because it would be buried deeper, at `project-name/stack-name/pulumi-pulumi-Stack/stack-name.yaml` instead of `project-name/stack-name/outputs.yaml`.

Alternatively we could just skip the type in the stack resource YAML path and always write it to `project-name/stack-name/stack.yaml`. But it still seems to me that it would be nice to have a dedicated file with just stack outputs.

WDYT?

pbzdyl replied:
What we have currently looks good to me and now I understand the rationale. Thanks!

pbzdyl commented on utilities/pulumi_state_splitter/tests/data.py:
```diff
@@ -0,0 +1,237 @@
+        "urn:pulumi:test-stack::test-project::pulumi:providers:provider::default"
+    )
+    provider_id = "deadbeef-dead-beef-dead-beefdeadbeef"
+    return util.resource_map(
```
In some places we are treating `pulumi:Stack` resource in a special way. Should it be included in the test resources for completeness?

filip-zyzniewski replied:
We do test that @ https://github.com/house-reliability-engineering/pulumi/blob/1ecd9a749fe7b0c2f505d6000a292cd46dc616c4/utilities/pulumi_state_splitter/tests/test_split.py#L147-L150, do we need more?

pbzdyl commented on utilities/pulumi_state_splitter/tests/test_cli.py:
```diff
@@ -0,0 +1,128 @@
+                    ".pulumi/stacks/test-project-2/test-stack-4.json",
+                ]
+            )
+        want = input_
```
I guess you wanted to run some assertions here instead of plain assignments.

filip-zyzniewski replied:
Thanks, fixed in 6ad6d00a1809635e202910980b261e5f56ac827f .

pbzdyl commented on utilities/pulumi_state_splitter/tests/test_stored_state.py:
```diff
@@ -0,0 +1,41 @@
+    import pulumi_state_splitter.stored_state
+
+
+class TestStackName(unittest.TestCase):
```
You might add a test case with invalid format that doesn't specify any `/` for completeness.

filip-zyzniewski replied:
Done in e84e7072745323c1878141f78ded91a3376c55ec .

pbzdyl commented on utilities/pulumi_state_splitter/tests/test_split.py:
```diff
@@ -0,0 +1,365 @@
+        input_.compare(got, self)
+
+    def test_unsplitter_new(self):
+        """Testing `Unsplitter` with with a new stack added."""
```
```suggestion
        """Testing `Unsplitter` with a new stack added."""
```

filip-zyzniewski replied:
Done in 7341de463e7e67935f8aafb6d238d35ed4bcf16d .

pbzdyl commented on utilities/pulumi_state_splitter/pulumi_state_splitter/split.py:
```diff
@@ -0,0 +1,177 @@
+            directory = cls.resource_subpath(resource.parent_resource).with_suffix("")
+        else:
+            directory = pathlib.Path()
+        return directory / resource.type / f"{resource.name}.yaml"
```
Notice that `resource.type` will include `:` that will make this file/directory name illegal on MS Windows.

filip-zyzniewski replied:
I wasn't planning on supporting windows, but done in 3bae12958a34e576ba5fe6081a460a016c0c6605 .

filip-zyzniewski replied:
And updated the unit test in 6c567c26976c945d7f4dffb698bab91e3c020eb4.

filip-zyzniewski committed:
pulumi_state_splitter: use actions/setup-python in the tests workflow

Per https://github.com/house-reliability-engineering/pulumi/pull/1/files#r1442576716

filip-zyzniewski committed:
pulumi_state_splitter: run GitHub Actions workflow when it changes

filip-zyzniewski committed:
pulumi_state_splitter: GitHub Actions workflow: set up poetry cache explicitly

instead of using the caching feature of action/setup-python, as it
needs to have the poetry binary in place:
https://github.com/actions/setup-python/issues/765
and I do not want to install poetry twice for two
Python versions.

This is to resolve
https://github.com/house-reliability-engineering/pulumi/pull/1/checks#step:3:14 :

```
 Error: Unable to locate executable file: poetry. Please verify either the file path exists or the file can be found within a directory specified by the PATH environment variable. Also check the file mode to verify the file is executable.
```

filip-zyzniewski committed:
pulumi_state_splitter: put project specific gitignore entries in its directory

Per
https://github.com/house-reliability-engineering/pulumi/pull/1#discussion_r1442573869

pbzdyl commented on .github/workflows/utilities-pulumi_state_splitter.yml:
```diff
@@ -0,0 +1,21 @@
+        with:
+            path: ~/.cache/pypoetry
+            key: ${{ hashFiles('utilities/pulumi_state_splitter/poetry.lock') }}
+      - run: ./utilities/pulumi_state_splitter/run-tests.sh
```
Nit pick: for better UX:
<img width="429" alt="image" src="https://github.com/house-reliability-engineering/pulumi/assets/57622/c8883b17-3c96-4c1d-aafe-fa781c3356ec">


```suggestion
      - name: Run tests
        run: ./utilities/pulumi_state_splitter/run-tests.sh
```

filip-zyzniewski replied:
Done in 853e7a798a26b83430cc9df760637019130d9e18 .

pbzdyl commented:
Tests fail but the workflow run is still green: https://github.com/house-reliability-engineering/pulumi/actions/runs/7423125176/job/20199847390?pr=1#step:5:23. Is it intended?

filip-zyzniewski committed:
pulumi_state_splitter.model: factor out _skip_some_falsy_values

Per
https://github.com/house-reliability-engineering/pulumi/pull/1/files/44e9d52405af04af95144efdb02c93421c38eee7#r1442676198

filip-zyzniewski committed:
pulumi_state_splitter.model.State: demand version 3

Per
https://github.com/house-reliability-engineering/pulumi/pull/1/files/#r1442680527

filip-zyzniewski committed:
pulumi_state_splitter: GitHub Actions workflow: name the Run tests step

filip-zyzniewski committed:
pulumi_state_splitter: test_style: sync black and isort line lengths

Otherwise they fight over import statements slightly shorted than 88
characters.

filip-zyzniewski committed:
pulumi_state_splitter: cli: process all stacks by default

Per
https://github.com/house-reliability-engineering/pulumi/pull/1#discussion_r1442973656

filip-zyzniewski committed:
pulumi_state_splitter: avoid colons in filenames

In case we ever got around to supporting Windows.

Per
https://github.com/house-reliability-engineering/pulumi/pull/1#discussion_r1442917922

filip-zyzniewski committed:
pulumi_state_splitter: cli test: add a missing assertion

Per
https://github.com/house-reliability-engineering/pulumi/pull/1/files#r1442851308

filip-zyzniewski committed:
pulumi_state_splitter: tests: a typo fix

Per
https://github.com/house-reliability-engineering/pulumi/pull/1/files#r1442902848

filip-zyzniewski committed:
pulumi_state_splitter: test_stored_state: test behavior for bad stack name components

Per https://github.com/house-reliability-engineering/pulumi/pull/1/files#r1442870909

filip-zyzniewski committed:
pulumi_state_splitter: test_split: test handling of colons in resource type names

Missed in 3bae12958a34e576ba5fe6081a460a016c0c6605 .

filip-zyzniewski committed:
pulumi_state_splitter: run-tests.sh: signal test failures with exit code

Per
https://github.com/house-reliability-engineering/pulumi/pull/1#issuecomment-1878794137

pbzdyl commented on .github/workflows/utilities-pulumi_state_splitter.yml:
```diff
@@ -0,0 +1,22 @@
+on:
```
You could also add a name for better UX:
<img width="826" alt="image" src="https://github.com/house-reliability-engineering/pulumi/assets/57622/267fe23a-2bff-4a4c-86c5-89d5e7c5e7df">

```suggestion
name: pulumi_state_splitter tests
on:
```

filip-zyzniewski replied:
WDYT about a46dec0e3220a5d84986e236d995190888ceeb34 ?

filip-zyzniewski committed:
pulumi_state_splitter: add a CLI smoke test

This would have caught the missing imports from
33812d405bdd39111e84db0d55764be5b3706ccb .

filip-zyzniewski committed:
pulumi_state_splitter: prevent a cyclic import

I have added too much in 33812d405bdd39111e84db0d55764be5b3706ccb .

This is to resolve pylint's:

```
tests/test_style.py:1:0: R0401: Cyclic import (pulumi_state_splitter -> pulumi_state_splitter.cli) (cyclic-import)
```

from
https://github.com/house-reliability-engineering/pulumi/actions/runs/7424886331/job/20205405823#step:5:23

filip-zyzniewski commented:
> Tests fail but the workflow run is still green: https://github.com/house-reliability-engineering/pulumi/actions/runs/7423125176/job/20199847390?pr=1#step:5:23. Is it intended?

Nope, fixed in 579f5ff8ebca1d60f848b29eaf591e75c486fe8f, and the test in b09cf4f98d2495567e2fe5d1013edc7fb826437e .

filip-zyzniewski committed:
pulumi_state_splitter: name the GitHub Actions workflow

Per
https://github.com/house-reliability-engineering/pulumi/pull/1#discussion_r1443187348

filip-zyzniewski committed:
pulumi_state_splitter: state_file: elaborate on resource sorting

Per
https://github.com/house-reliability-engineering/pulumi/pull/1#discussion_r1442694627

filip-zyzniewski committed:
pulumi_state_splitter: split: save no outputs as {}

So that the type stays the same.

pbzdyl commented on .github/workflows/utilities-pulumi_state_splitter.yml:
```diff
@@ -0,0 +1,24 @@
+        with:
+          python-version: '3.11'
+      - uses: actions/cache@v3
+        with:
+            path: ~/.cache/pypoetry
+            key: ${{ hashFiles('utilities/pulumi_state_splitter/poetry.lock') }}
```
Indentation is off:
```suggestion
        with:
          path: ~/.cache/pypoetry
          key: ${{ hashFiles('utilities/pulumi_state_splitter/poetry.lock') }}
```

filip-zyzniewski replied:
Done in cb4d7925f71be2487ea8ddb93d28130d2c1064e2 .

filip-zyzniewski committed:
pulumi_state_splitter: GitHub Actions workflow: fix indentation

Per
https://github.com/house-reliability-engineering/pulumi/pull/1/files#r1444253928
